src/index.js

```diff
@@ -4,224 +4,167 @@
+  if (! (type in this.listeners)) {
+    return;
   }
-
-  // These are the three public API methods that
-  // consumers can use to connunicate over TLS.
-
-  async send(data) {
-    assertIsBytes(data);
-    await this._synchronized(async () => {
-      await this._state.sendApplicationData(data);
-    });
+  var stack = this.listeners[type];
+  for (var i = 0, l = stack.length; i < l; i++) {
+    if (stack[i] === callback) {
+      stack.splice(i, 1);
+      return;
+    }
   }
+};
 
-  async recv(data) {
-    assertIsBytes(data);
-    return await this._synchronized(async () => {
-      const [type, buf] = await this._recordReceiver.recv(data);
-      return await this._dispatchIncomingRecord(type, buf);
-    });
+EventTarget.prototype.dispatchEvent = function(event) {
+  if (! (event.type in this.listeners)) {
+    return true;
   }
+  var stack = this.listeners[event.type].slice();
 
-  async close() {
-    await this._synchronized(async () => {
-      await this._state.close();
-    });
+  for (var i = 0, l = stack.length; i < l; i++) {
+    stack[i].call(this, event);
   }
-
-  // Ensure that async functions execute one at a time,
-  // by waiting for the previous call to `_synchronized()` to complete
-  // before starting a new one.  This helps ensure that we complete
-  // one state-machine transition before starting to do the next.
-
-  _synchronized(cb) {
-    const nextPromise = this._lastPromise.then(() => {
-      return cb();
-    }).catch(async err => {
-      // All errors immediately put us in a terminal "error" state.
-      await this._transition(STATE.ERROR, err);
-      throw err;
-    });
-    // We don't want to hold on to the return value or error,
-    // just synchronize on the fact that it completed.
-    this._lastPromise = nextPromise.then(noop, noop);
-    return nextPromise;
+  return ! event.defaultPrevented;
+};
+
+export class InsecurePairingChannel extends EventTarget {
+  constructor(channelId, channelKey, socket, tlsConnection) {
+    super();
+    this._channelId = channelId;
+    this._channelKey = channelKey;
+    this._socket = socket;
+    this._tlsConnection = tlsConnection;
+    this._setupListeners();
   }
 
-  // This drives internal transition of the state-machine,
-  // ensuring that the new state is properly initialized.
-
-  async _transition(State, ...args) {
-    this._state = new State(this);
-    await this._state.initialize(...args);
+  /**
+   * Create a new pairing channel.
+   *
+   * @returns Promise<InsecurePairingChannel>
+   */
+  static create(channelServerURI, win = window) {
+    const channelKey = win.crypto.getRandomValues(new Uint8Array(32));
+    return this._makePairingChannel(win, channelServerURI, InsecureServerConnection, channelKey);
   }
 
-  // These are helpers to allow the state to add data to the next outgoing record.
+  /**
+   * Connect to an existing pairing channel.
+   *
+   * `channelServerURI` must be the full URI and therefore should contain the channel ID.
```

Comment:
I'm very lightly -1 on making the caller construct a channelserver URI to pass in here, as it means that the caller needs to know about the semantics of the channelserver URI space.  Would it work just as well to pass in the base channelserverURI and the channelId as two separate arguments, and keep all the knowledge of how channelserver works contained within the `PairingChannel` class?